### ISSUE TITLE
[Bug]: useParameter() returns nothing for a docs page

### ISSUE THREAD

**Issue by probablyup** [storybookjs/storybook#21798]
### Describe the bug

I'm attempting to upgrade a storybook 6 project to storybook 7 and am running into an issue where the `storybook-dark-mode` doesn't work for docs pages. 

I've narrowed down the issue to `@storybook/api` `useParameter('darkMode')` returning an empty object inside the addon, but _only_ for rendering docs pages.

### To Reproduce

I don't have a reproduction at the moment, but a basic storybook v7 project with `storybook-dark-mode` added should exhibit the behavior. In chrome devtools I looked up the `manager.js` file from storybook-dark-mode and added a breakpoint where it uses the `useParameter` hook.

My guess is the parameters store isn't getting initialized for docs pages for some reason?

**Comment by tmeasday** [comment]
Hey @probablyup, thanks for calling this out. We've taken another look at this.

The root of the problem is that with the new docs architecture, when you view docs there isn't an underlying story. In 6.5 there always was (although it was a bit of a kludge in the case of .mdx files). What that meant in 6.5 was that you could always get parameters when visiting a docs page, because there'd always be a story to read them off. In the case of .mdx files, that was just the project-level parameters (preview.js), which happened to work great for this addon.

In any case, we did look at this problem when migrated some of our own addons but ultimately decided in the cases where they needed useParameter() it didn't really make sense to even do that on docs pages, and so we thought the changed behaviour was probably the best way forward.

Having seen this example, it does seem like there are/will be other addons that do need the project-level parameters to be available all the time. So we'll look at making it work again on docs pages.

As we are super late in the 7.0 release cycle, we need to be a bit careful about how we do this. We'll likely release the new mechanism in a 7.1 alpha in the near future, test it a little, and backport it as a bug fix in 7.0.x patch release.

**Comment by tmeasday** [comment]
That's a limitation of the "on demand" nature of the store. We can't know the parameters of a story until it has been rendered

**Comment by probablyup** [comment]
Can’t you use what’s provided in preview.js as a global default?

**Comment by tmeasday** [comment]
@probablyup yes, in fact as of https://github.com/storybookjs/storybook/pull/22071 we'd supply the component's parameters (ie. what's defined in the default export of the CSF file + the project's parameters defined in `preview.js`) in this case. 

Would that be a good or a bad thing for your use case @raphaelokon? So your call of `storybookAPI.getParameters(item.id)` would return:
 - the project's parameters before you render any story for the component
 - the component's parameters before you render story in question (if you render a sibling)
 - the story's parameters once you've rendered the story.

The inconsistency doesn't sit super well with me but maybe it's better than nothing as @probablyup implies.

As an aside the issue of not getting a sibling story's parameters is an issue in the v7 (on demand) store prior to 7.0, IIUC.

**Comment by tmeasday** [comment]
@raphaelokon no problem at all, in fact it was helpful to think about this use case! I was just noting it because the focus here is to fix the regression in 7.0, rather than necessarily solve every problem with `useParameter()`.

**Comment by tmeasday** [comment]
@raphaelokon so after some discussion with @shilman, we are going to take a simpler approach in https://github.com/storybookjs/storybook/pull/22118.

What this will mean is that `getParameters()` will still display the current behaviour for stories that haven't been rendered yet (that is returning nothing).

There seem to be two clear use cases for knowing the project/component parameters on the manager side:

1. To fallback if we don't yet know the story's parameters (your use case).
2. To fallback as we wait for the story to prepare (https://github.com/storybookjs/storybook/issues/22119). In this case it isn't clear that falling back is even the correct solution, would love to hear people's thoughts.

So regarding 1. I'd like to hear more about your use case. Which parameters do you need exactly to `renderLabel`? Do they change between components/stories?

**Comment by probablyup** [comment]
I’d really like to avoid returning nothing during render to avoid FOUC inside the preview window. It would make the dark mode addon have imperfect UX since it couldn’t proactively set dark mode based on config in preview.js. 

Really would just love parameters exported from preview to be the default, it would solve all problems I’m aware of.

**Comment by tmeasday** [comment]
@probablyup can you comment on https://github.com/storybookjs/storybook/issues/22119 with regards to that? I'm interested in both the initial FOUC and the FOUC as you change stories, which potentially could be handled differently.

For SB dark mode it's unlikely you'd set the parameters differently per story or per component, but that's not the only addon that calls `useParameters()` and changing the behaviour of other addons to fallback to the project parameters may have unexpected issues, which is why we are being cautious here late in the 7.0 cycle.

In terms of https://github.com/storybookjs/storybook/pull/22118, AFAIK it restores the 6.5 behaviour so is safe to merge into 7.0 as a patch release.

**Comment by shilman** [comment]
Crikey!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.1.0-alpha.6 containing PR #22120 that references this issue. Upgrade today to the `@future` NPM tag to try it out!

```
npx sb@next upgrade --tag future
```

**Comment by tmeasday** [comment]
> Regrading the second question, no, they are just static strings in our use case.

I meant are they different per-story or component, which it seems like they are.

We actually have the perfect mechanism for this, tags. It's not a properly released feature yet, but we already use it for autodocs: https://storybook.js.org/docs/react/writing-docs/autodocs#setup-automated-documentation

Tags are different from parameters as they are static and part of the index, so they are available immediately as soon as the menu is rendered. I think the tags are available on the `item` passed to `renderLabel()` already!

**Comment by tmeasday** [comment]
@raphaelokon would it make sense to use a field in `main.js` to configure that? Something like:

```ts
addons: [{ name: 'your addon', options: { x: 'y' } }]
```

I don't believe it's currently simply to get those options in the manager, but we could make it so.

**Comment by tmeasday** [comment]
Oh, what kind of per story data would you use?

**Comment by tmeasday** [comment]
> This could contain even more related data to this specific status

Would this specific information be relevant to the component/story in question?

**Comment by shilman** [comment]
¡Ay Caramba!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.1.0-alpha.8 containing PR #22154 that references this issue. Upgrade today to the `@future` NPM tag to try it out!

```
npx sb@next upgrade --tag future
```

**Comment by shilman** [comment]
Jeepers creepers!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.0.7 containing PR #22120 that references this issue. Upgrade today to the `@latest` NPM tag to try it out!

```
npx sb@latest upgrade
```

**Comment by tmeasday** [comment]
@probablyup I don't believe any changes should be necessary.

I tried it out on my repro repo with 7.0.7 and the v6 store, and it appears to work as I expected. What am I missing?

https://github.com/tmeasday/test-dark-mode/tree/dark-mode-7-ssv6

**Comment by probablyup** [comment]
Sigh, I don’t know what the deal is then. I’ve been very on top of upgrading these libraries and honestly it’s been a misery.

**Comment by tmeasday** [comment]
What (mis-) behaviour are you seeing?

**Comment by probablyup** [comment]
I upgraded to 7.0.7 and ran yarn dedupe to ensure everything is on the proper version, verified the lockfile etc. dark mode addon still is showing empty for the useParameter hook on a docs page

**Comment by tmeasday** [comment]
Hmmm, that's pretty weird. I'm not sure what to suggest there, can you turn on `logLevel: 'debug'` and check the console messages for the `setStories` event?

**Comment by probablyup** [comment]
Will try that and report back thanks!

**Comment by shilman** [comment]
Ta-da!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.0.8 containing PR #22154 that references this issue. Upgrade today to the `@latest` NPM tag to try it out!

```
npx sb@latest upgrade
```

**Comment by tmeasday** [comment]
Hold on, does that comment mean that 7.0.7 didn't have the fix for the v6 store? That's kind of at odds with the screenshot above 🤔

**Comment by shilman** [comment]
@tmeasday my bad. Will fix by hand and hopefully the new release workflow helps

**Comment by nareshbhatia** [comment]
Hard to follow how to get this fix. I am at v7.0.11 and still seeing the issue. Do I have to wait till 7.1.0 gets released?

**Comment by tmeasday** [comment]
@nareshbhatia the fix should be in 7.0.7. What is the behaviour you are (not) seeing?

**Comment by nareshbhatia** [comment]
@tmeasday, I am using `storybook-dark-mode@3.0.0` with `storybook@7.0.11`. When I turn on dark mode, only the storybook UI changes to dark mode, the story itself retains the white background. See below:

<img width="677" alt="image" src="https://github.com/storybookjs/storybook/assets/1220198/d27fd859-3ca4-4621-a037-2a5a6bc3dd2c">

If I look at the markup in chrome debugger, the `theme-dark` class is correctly applied to the `<html>` element, but `<body>` class does not get a dark background:

<img width="323" alt="image" src="https://github.com/storybookjs/storybook/assets/1220198/d8d6325b-2671-4d05-9f77-b148c1317711">

**Comment by tmeasday** [comment]
@nareshbhatia hmm, that sounds like a different issue than what this ticket is about [1].

I'm not sure about the behaviour you are seeing. If you can post a reproduction that shows it working 6.5 and broken in 7.0 I can take a look, but maybe in a new ticket?


[1] this ticket is about (a) docs entries and (b) the manager side, ie. the sidebar etc.

**Comment by nareshbhatia** [comment]
Sounds good, @tmeasday. I will create a reproduction and a new issue.

**Comment by shubhamdhingra007** [comment]
@nareshbhatia, did you find any resolution for this?

**Comment by nareshbhatia** [comment]
@shubhamdhingra007, it's still on my TODO list. Haven't gotten to it yet.
